### PR TITLE
Kusto Integration

### PR DESCRIPTION
**Pull request by barnuri**

### CODE REVIEW

**Comment by ErikEJ** [comment]
Looks like you miss some "where is null" checks?

File: src/Core/RevEng.Core.60/Routines/Functions/SqlServerFunctionModelFactory.cs [hunk 1]
```diff
@@ -21,19 +21,10 @@ public SqlServerFunctionModelFactory()
     ROUTINE_NAME,
     CAST(CASE WHEN (DATA_TYPE != 'TABLE') THEN 1 ELSE 0 END AS bit) AS IS_SCALAR
 FROM INFORMATION_SCHEMA.ROUTINES
+LEFT JOIN sys.extended_properties AS [ep] on [ep].major_id = object_id(QUOTENAME(ROUTINE_SCHEMA) + '.' + QUOTENAME(ROUTINE_NAME)) AND [ep].minor_id = 0 AND [ep].class = 1
```
Comment (ErikEJ) [inline src/Core/RevEng.Core.60/Routines/Functions/SqlServerFunctionModelFactory.cs:24]: Where is the name criteria??

File: src/Core/RevEng.Core.60/Routines/Procedures/SqlServerStoredProcedureModelFactory.cs [hunk 2]
```diff
@@ -23,19 +23,10 @@ public SqlServerStoredProcedureModelFactory()
     ROUTINE_NAME,
     CAST(0 AS bit) AS IS_SCALAR
 FROM INFORMATION_SCHEMA.ROUTINES
+LEFT JOIN sys.extended_properties AS [ep] on [ep].major_id = object_id(QUOTENAME(ROUTINE_SCHEMA) + '.' + QUOTENAME(ROUTINE_NAME)) AND [ep].minor_id = 0 AND [ep].class = 1
```
Comment (ErikEJ) [inline src/Core/RevEng.Core.60/Routines/Procedures/SqlServerStoredProcedureModelFactory.cs:26]: name criteria?

File: src/Core/RevEng.Core.80/PatchedSqlServerDatabaseModelFactory.cs [hunk 3]
```diff
@@ -748,6 +743,7 @@ private void GetColumns(
 (
     SELECT [t].[name], [t].[object_id], [t].[schema_id]
     FROM [sys].[tables] t
+    LEFT JOIN [sys].[extended_properties] AS [ep] ON [ep].[major_id] = [t].[object_id] AND [ep].[minor_id] = 0 AND [ep].[class] = 1 AND [ep].[name] = 'microsoft_database_tools_support'
```
Comment (ErikEJ) [inline src/Core/RevEng.Core.80/PatchedSqlServerDatabaseModelFactory.cs:746]: Where criteria missing

File: src/Core/RevEng.Core.80/PatchedSqlServerDatabaseModelFactory.cs [hunk 4]
```diff
@@ -1049,6 +1045,7 @@ FROM [sys].[indexes] AS [i]
 JOIN [sys].[tables] AS [t] ON [i].[object_id] = [t].[object_id]
 JOIN [sys].[index_columns] AS [ic] ON [i].[object_id] = [ic].[object_id] AND [i].[index_id] = [ic].[index_id]
 JOIN [sys].[columns] AS [c] ON [ic].[object_id] = [c].[object_id] AND [ic].[column_id] = [c].[column_id]
+LEFT JOIN [sys].[extended_properties] AS [ep] ON [ep].[major_id] = [t].[object_id] AND [ep].[minor_id] = 0 AND [ep].[class] = 1 AND [ep].[name] = 'microsoft_database_tools_support'
```
Comment (ErikEJ) [inline src/Core/RevEng.Core.80/PatchedSqlServerDatabaseModelFactory.cs:1048]: Where criteria missing

File: src/Core/RevEng.Core.80/PatchedSqlServerDatabaseModelFactory.cs (hunk 3 above)
Comment (ErikEJ) [inline src/Core/RevEng.Core.80/PatchedSqlServerDatabaseModelFactory.cs:746]: Why was this added here?

File: src/Core/RevEng.Core.80/PatchedSqlServerDatabaseModelFactory.cs [hunk 5]
```diff
@@ -1492,6 +1491,7 @@ private void GetTriggers(DbConnection connection, IReadOnlyList<DatabaseTable> t
     [tr].[name] AS [trigger_name]
 FROM [sys].[triggers] AS [tr]
 JOIN [sys].[tables] AS [t] ON [tr].[parent_id] = [t].[object_id]
+LEFT JOIN [sys].[extended_properties] AS [ep] ON [ep].[major_id] = [t].[object_id] AND [ep].[minor_id] = 0 AND [ep].[class] = 1 AND [ep].[name] = 'microsoft_database_tools_support'
```
Comment (ErikEJ) [inline src/Core/RevEng.Core.80/PatchedSqlServerDatabaseModelFactory.cs:1494]: Why was this added?

File: src/Core/RevEng.Core.80/PatchedSqlServerDatabaseModelFactory.cs [hunk 6]
```diff
@@ -1063,6 +1060,7 @@ FROM [sys].[indexes] i
    JOIN [sys].[tables] AS [t] ON [i].[object_id] = [t].[object_id]
    JOIN [sys].[index_columns] AS [ic] ON [i].[object_id] = [ic].[object_id] AND [i].[index_id] = [ic].[index_id]
    JOIN [sys].[columns] AS [c] ON [ic].[object_id] = [c].[object_id] AND [ic].[column_id] = [c].[column_id]
+   LEFT JOIN [sys].[extended_properties] AS [ep] ON [ep].[major_id] = [t].[object_id] AND [ep].[minor_id] = 0 AND [ep].[class] = 1 AND [ep].[name] = 'microsoft_database_tools_support'
```
Comment (ErikEJ) [inline src/Core/RevEng.Core.80/PatchedSqlServerDatabaseModelFactory.cs:1063]: Why was this added here?

**Review by ErikEJ — CHANGES_REQUESTED** [review]
Some updates seem  to be needed

File: src/Core/RevEng.Core.80/PatchedSqlServerDatabaseModelFactory.cs [hunk 7]
```diff
@@ -1310,6 +1308,7 @@ FROM [sys].[foreign_keys] AS [f]
 JOIN [sys].[columns] AS [col1] ON [col1].[column_id] = [fc].[parent_column_id] AND [col1].[object_id] = [t].[object_id]
 JOIN [sys].[tables] AS [tab2] ON [tab2].[object_id] = [fc].[referenced_object_id]
 JOIN [sys].[columns] AS [col2] ON [col2].[column_id] = [fc].[referenced_column_id] AND [col2].[object_id] = [tab2].[object_id]
+LEFT JOIN [sys].[extended_properties] AS [ep] ON [ep].[major_id] = [t].[object_id] AND [ep].[minor_id] = 0 AND [ep].[class] = 1 AND [ep].[name] = 'microsoft_database_tools_support'
```
[the remark this answers is not in the document]
Comment (barnuri) [reply src/Core/RevEng.Core.80/PatchedSqlServerDatabaseModelFactory.cs:1311]: because of the table filter that added anywhere

File: src/Core/RevEng.Core.80/PatchedSqlServerDatabaseModelFactory.cs (hunk 5 above)
Comment (barnuri) [reply src/Core/RevEng.Core.80/PatchedSqlServerDatabaseModelFactory.cs:1494]: because of the table filter that added anywhere

File: src/Core/RevEng.Core.80/PatchedSqlServerDatabaseModelFactory.cs (hunk 6 above)
Comment (barnuri) [reply src/Core/RevEng.Core.80/PatchedSqlServerDatabaseModelFactory.cs:1063]: because of the table filter that added anywhere

File: src/Core/RevEng.Core.80/PatchedSqlServerDatabaseModelFactory.cs (hunk 3 above)
Comment (barnuri) [reply src/Core/RevEng.Core.80/PatchedSqlServerDatabaseModelFactory.cs:746]: because of the table filter that added anywhere

File: src/Core/RevEng.Core.60/Routines/Functions/SqlServerFunctionModelFactory.cs (hunk 1 above)
Comment (barnuri) [reply src/Core/RevEng.Core.60/Routines/Functions/SqlServerFunctionModelFactory.cs:24]: what do you mean ?

Comment (barnuri) [reply src/Core/RevEng.Core.60/Routines/Functions/SqlServerFunctionModelFactory.cs:24]: wrap it with [] ?

Comment (barnuri) [reply src/Core/RevEng.Core.60/Routines/Functions/SqlServerFunctionModelFactory.cs:24]: wanted to keep in same style like the other table

Comment (ErikEJ) [reply src/Core/RevEng.Core.60/Routines/Functions/SqlServerFunctionModelFactory.cs:24]: I mean the name = 'microsoft_toolsxxx' criteria

File: src/Core/RevEng.Core.60/Routines/Procedures/SqlServerStoredProcedureModelFactory.cs (hunk 2 above)
Comment (barnuri) [reply src/Core/RevEng.Core.60/Routines/Procedures/SqlServerStoredProcedureModelFactory.cs]: fixed

File: src/Core/RevEng.Core.60/Routines/Functions/SqlServerFunctionModelFactory.cs (hunk 1 above)
Comment (barnuri) [reply src/Core/RevEng.Core.60/Routines/Functions/SqlServerFunctionModelFactory.cs]: fixed

File: src/Core/RevEng.Core.60/Routines/Functions/SqlServerFunctionModelFactory.cs [hunk 8]
```diff
@@ -21,19 +21,10 @@ public SqlServerFunctionModelFactory()
     ROUTINE_NAME,
     CAST(CASE WHEN (DATA_TYPE != 'TABLE') THEN 1 ELSE 0 END AS bit) AS IS_SCALAR
 FROM INFORMATION_SCHEMA.ROUTINES
+LEFT JOIN sys.extended_properties AS [ep] on [ep].major_id = object_id(QUOTENAME(ROUTINE_SCHEMA) + '.' + QUOTENAME(ROUTINE_NAME)) AND [ep].minor_id = 0 AND [ep].class = 1 and [ep].name = N'microsoft_database_tools_support'
```
Comment (ErikEJ) [inline src/Core/RevEng.Core.60/Routines/Functions/SqlServerFunctionModelFactory.cs:24]: This is not needed for Functions

File: src/Core/RevEng.Core.60/Routines/Procedures/SqlServerStoredProcedureModelFactory.cs [hunk 9]
```diff
@@ -23,19 +23,10 @@ public SqlServerStoredProcedureModelFactory()
     ROUTINE_NAME,
     CAST(0 AS bit) AS IS_SCALAR
 FROM INFORMATION_SCHEMA.ROUTINES
+LEFT JOIN sys.extended_properties AS [ep] on [ep].major_id = object_id(QUOTENAME(ROUTINE_SCHEMA) + '.' + QUOTENAME(ROUTINE_NAME)) AND [ep].minor_id = 0 AND [ep].class = 1 and [ep].name = N'microsoft_database_tools_support'
```
Comment (ErikEJ) [inline src/Core/RevEng.Core.60/Routines/Procedures/SqlServerStoredProcedureModelFactory.cs:26]: This is not needed for stored procedures

File: src/Core/RevEng.Core.60/Routines/Functions/SqlServerFunctionModelFactory.cs (hunk 8 above)
Comment (barnuri) [reply src/Core/RevEng.Core.60/Routines/Functions/SqlServerFunctionModelFactory.cs:24]: are you sure ? because it was before as subquery

Comment (ErikEJ) [reply src/Core/RevEng.Core.60/Routines/Functions/SqlServerFunctionModelFactory.cs:24]: Sorry, misread the changes

**Comment by ErikEJ** [comment]
Looking good, what are the test errors on the EF Core PR?

**Comment by barnuri** [comment]
> Looking good, what are the test errors on the EF Core PR?

im not sure but this pr works e2e
yesterday i saw one failure because missing cp from this pr
but now im not sure what it is can you take a look ?

**Comment by ErikEJ** [comment]
@barnuri Just had a look, it looks very unrelated.

**Review by ErikEJ — APPROVED** [review]

**Comment by ErikEJ** [comment]
@barnuri Thanks a lot for this, looking forward to hearin about your experience

**Comment by barnuri** [comment]
> @barnuri Thanks a lot for this, looking forward to hearin about your experience

Thanks it was nice to work with you 😄
can we trigger now a new release? 
or i need to wait for nightly build

**Comment by ErikEJ** [comment]
The Nightly build is being published right now
